### PR TITLE
Stage and deploy apt packages on S3

### PR DESCRIPTION
**Pull request by umanwizard**

### CODE REVIEW

File: ci/deploy/linux.py [hunk 1]
```diff
@@ -58,6 +61,32 @@ def publish_deb(package: str, version: str) -> None:
...
 
+    s3 = boto3.client("s3")
+    bucket = "materialize-apt-repository"
+    object_key = f"pool/stable/m/ma/materialized-{version}.deb"
+    # Download the staged package (deb-s3 needs to get various metadata from it)
+    s3.download_file(bucket, object_key, f"materialized-{version}.deb")
+    # Import the private key into GPG
+    key = os.environ["GPG_KEY"].replace("\\n", "\n")
+    gpg = subprocess.Popen(["gpg", "--import"], stdin=subprocess.PIPE)
+    gpg.communicate(key.encode("ascii"))
```
Comment (benesch) [inline ci/deploy/linux.py:72]: `Popen` is roughly deprecated in favor of `subprocess.run`, though in this case I think you can just use `spawn.runv`?

```python
spawn.runv(["gpg", "--import"], stdin=io.BytesIO(os.environ["GPG_KEY"]))
```

File: ci/deploy/linux.py [hunk 2]
```diff
@@ -58,6 +61,32 @@ def publish_deb(package: str, version: str) -> None:
     )
     bt.repo("apt").package(package).publish_uploads(version)
 
+    s3 = boto3.client("s3")
+    bucket = "materialize-apt-repository"
+    object_key = f"pool/stable/m/ma/materialized-{version}.deb"
+    # Download the staged package (deb-s3 needs to get various metadata from it)
+    s3.download_file(bucket, object_key, f"materialized-{version}.deb")
+    # Import the private key into GPG
+    key = os.environ["GPG_KEY"].replace("\\n", "\n")
```
Comment (benesch) [inline ci/deploy/linux.py:70]: This seems funky. Can't we just make the environment variable contain the newlines directly?

**Review by benesch — APPROVED** [review]

File: ci/deploy/linux.py (hunk 1 above)
Comment (umanwizard) [reply ci/deploy/linux.py:72]: No, you can't use BytesIO/StringIO as input to subprocess.run .

```
>>> subprocess.run(["cat"], stdin=io.BytesIO(b"test"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 505, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 829, in __init__
    errread, errwrite) = self._get_handles(stdin, stdout, stderr)
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 1586, in _get_handles
    p2cread = stdin.fileno()
io.UnsupportedOperation: fileno
```

File: ci/deploy/linux.py (hunk 2 above)
Comment (umanwizard) [reply ci/deploy/linux.py:70]: Perhaps, but I don't know how.

```
bash-3.2$ GPG_KEY=foo\
> bar\
> quux
bash-3.2$ echo $GPG_KEY
foobarquux
bash-3.2$ GPK_KEY="foo
> bar
> quux"
bash-3.2$ echo $GPG_KEY
foobarquux
```



Might just be that I'm lacking some basic knowledge of shell escaping.

Comment (benesch) [reply ci/deploy/linux.py:70]: The last example is legit; you just need to be sure to use quotes when you print the value of the env var in bash or you won't see the embedded newlines:

```bash
GPG_KEY="foo
bar
queue"
echo "$GPG_KEY"
```

File: ci/deploy/linux.py (hunk 1 above)
Comment (benesch) [reply ci/deploy/linux.py:72]: Ugh. Ok. Sorry to mislead you.

Comment (umanwizard) [reply ci/deploy/linux.py:72]: It's fine, that was the first thing I tried also

File: ci/deploy/linux.py (hunk 2 above)
Comment (umanwizard) [reply ci/deploy/linux.py:70]:
```
bash-3.2$ GPK_KEY="foo
bar
quux"
bash-3.2$ echo "$GPG_KEY"
foobarquux
```

Comment (benesch) [reply ci/deploy/linux.py:70]: You typo’d “GPG_KEY” there.

Comment (umanwizard) [reply ci/deploy/linux.py:70]: There are bashisms that apparently let you do this, like `printf -v`, but I prefer to keep shell stuff POSIX-compliant unless necessary. No strong opinion though really.

Comment (umanwizard) [reply ci/deploy/linux.py:70]: wow, I am stupid

**Comment by benesch** [comment]
:D

**Review by benesch — APPROVED** [review]

**Pull request merged**